lib/hammer_cli/output/field_filter.rb

```diff
@@ -1,21 +1,87 @@
 module HammerCLI::Output
-
   class FieldFilter
+    attr_reader :fields, :filtered_fields
+    attr_accessor :classes_filter, :sets_filter
+
+    def initialize(fields = [], filters = {})
+      self.fields = fields
+      @classes_filter = filters[:classes_filter] || []
+      @sets_filter = filters[:sets_filter] || []
+    end
 
-    def initialize(field_classes=[])
-      @field_classes = field_classes
+    def fields=(fields)
+      @fields = fields || []
+      @filtered_fields = @fields.dup
     end
 
-    def filter(fields)
-      fields = fields.clone
-      @field_classes.each do |cls|
-        fields.reject! do |f|
+    def filter_by_classes(classes = nil)
+      classes ||= @classes_filter
+      classes.each do |cls|
+        @filtered_fields.reject! do |f|
           f.is_a? cls
         end
       end
-      fields
+      self
     end
 
-  end
+    def filter_by_sets(sets = nil)
+      sets ||= @sets_filter
+      return self if sets.empty?
+
+      set_names, labels = resolve_set_names(sets)
+      deep_filter(@filtered_fields, set_names, labels)
+      self
+    end
+
+    def filter_by_data(data, compact_only: false)
+      @filtered_fields = displayable_fields(@filtered_fields,
+                                            data,
+                                            compact_only: compact_only)
+      self
+    end
+
+    private
+
+    def deep_filter(fields, set_names, labels)
+      fields.select! do |f|
+        allowed = include_by_label?(labels, f.full_label.downcase)
+        allowed ||= (f.sets & set_names).any?
+        deep_filter(f.fields, set_names, labels) if f.respond_to?(:fields)
+        allowed
+      end
+    end
 
+    def displayable_fields(fields, record, compact_only: false)
+      fields.select do |field|
+        field_data = HammerCLI::Output::Adapter::Abstract.data_for_field(
+          field, record
+        )
+        if compact_only && !field_data.is_a?(HammerCLI::Output::DataMissing)
+          true
+        else
+          field.display?(field_data)
+        end
+      end
+    end
+
+    def include_by_label?(labels, label)
+      return true if labels.include?(label)
+
+      labels.each do |l|
+        return true if l.start_with?("#{label}/") || label.start_with?(l)
```

Comment:
@ofedoren finally I was able to figure out how this is supposed to work and fixed the condition to autocomplete with asterisk `l.start_with?("#{label}/") || label.match(%r{^#{l.gsub(/\*/, '.*')}(|\/.*)$})`. I simplified it a bit also and have
```ruby
   def include_by_label?(labels, label)
      labels.any? do |l|
        l.start_with?("#{label}/") || label.match(%r{^#{l.gsub(/\*/, '.*')}(|\/.*)$})
      end
    end
```
Now it works with:
- `host info --id 4 --fields 'id,name,network'`
- `host info --id 4 --fields 'id,name,network*'`
- `host info --id 4 --fields 'id,name,network/domain'`

What do you think?